File: HOT100/13.堆/295.数据流的中位数.cpp
/*
 * @lc app=leetcode.cn id=295 lang=cpp
 *
 * [295] 数据流的中位数
 */
#include "tools.h"
// @lc code=start
class MedianFinder {
public:
    priority_queue<int> left;
    priority_queue<int, vector<int>, greater<>> right;
    MedianFinder() {
        
    }
    
    void addNum(int num) {
        if (left.size() == right.size())
        {
            right.push(num);
            left.push(right.top());
            right.pop();
        }
        else
        {
            left.push(num);
            right.push(left.top());
            left.pop();
        }
    }
    
    double findMedian() {
        if (left.size() > right.size()) return left.top();
        else return (left.top() + right.top()) / 2.0f;
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */
// @lc code=end

File: 题单/&/295.数据流的中位数.py
#
# @lc app=leetcode.cn id=295 lang=python3
#
# [295] 数据流的中位数
#
from mytools import *
# @lc code=start
class MedianFinder:

    def __init__(self):
        self.l = []
        self.r = []

    def addNum(self, num: int) -> None:
        heappush(self.l, -num)
        heappush(self.r, -heappop(self.l))
        if len(self.r) > len(self.l):
            heappush(self.l, -heappop(self.r))

    def findMedian(self) -> float:
        n = len(self.l) + len(self.r)
        if n & 1: return -self.l[0]
        else: return (self.r[0] - self.l[0]) / 2


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()
# @lc code=end

